discord/http.py

```diff
@@ -260,123 +235,135 @@ async def request(
+                        ) as response:
                             _log.debug(
-                                (
-                                    "A rate limit bucket has been exhausted (bucket:"
-                                    " %s, retry: %s)."
-                                ),
-                                bucket,
-                                delta,
+                                "%s %s with %s has returned %s",
+                                method,
+                                url,
+                                kwargs.get("data"),
+                                response.status,
                             )
-                            maybe_lock.defer()
-                            self.loop.call_later(delta, lock.release)
-
-                        # the request was successful so just return the text/json
-                        if 300 > response.status >= 200:
-                            _log.debug("%s %s has received %s", method, url, data)
-                            return data
-
-                        # we are being rate limited
-                        if response.status == 429:
-                            if not response.headers.get("Via") or isinstance(data, str):
-                                # Banned by Cloudflare more than likely.
-                                raise HTTPException(response, data)
 
-                            fmt = (
-                                "We are being rate limited. Retrying in %.2f seconds."
-                                ' Handled under the bucket "%s"'
+                            # even errors have text involved in them so this is safe to call
+                            data = await json_or_text(response)
+
+                            # check if we have rate limit header information
+                            remaining = response.headers.get("X-Ratelimit-Remaining")
+                            reset_after = response.headers.get(
+                                "X-Ratelimit-Reset-After"
                             )
 
-                            # sleep a bit
-                            retry_after: float = data["retry_after"]
-                            _log.warning(fmt, retry_after, bucket)
-
-                            # check if it's a global rate limit
-                            is_global = data.get("global", False)
-                            if is_global:
-                                _log.warning(
-                                    (
-                                        "Global rate limit has been hit. Retrying in"
-                                        " %.2f seconds."
-                                    ),
-                                    retry_after,
+                            if remaining:
+                                remaining = int(remaining)
+                            if reset_after:
+                                reset_after = float(reset_after)
```

Comment:
This doesn't take `assume_unsync_clock` (which is a parameter on the client that controls whether `X-RateLimit-Reset-After` or `X-RateLimit-Reset` subtracted by the current time is used) into account.

Reply:
why would reset after need to be subtracted from the current time being used? it is a value in seconds, not a datetime.

Reply:
Well it's just an option on the client to rely on the system's clock instead of Discord. I think it would be good to continue supporting this instead of silently dropping support. Also, there's a utility function named `_parse_ratelimit_headers` that can either return `X-RateLimit-Reset-After` or the seconds of the difference of `X-RateLimit-Reset` from the current system time.